Allow configuring maximum document ID length during replication

nickva opened the pull request:
Currently due to a bug in http parser and lack of document ID length
enforcement, large document IDs will break replication jobs. Large IDs
will pass through the _change feed, revs diffs,  but then fail
during open_revs get request. open_revs request will keep retrying until
it gives up after long enough time, then replication task crashes and
restart again with the same pattern. The current effective limit is
around 8k or so. (The buffer size default 8192 and if the first line
of the request is larger than that, request will fail).

(See http://erlang.org/pipermail/erlang-questions/2011-June/059567.html
for more information about the possible failure mechanism).

Bypassing the parser bug by increasing recbuf size, will alow replication
to finish, however that means simply spreading the abnormal document through
the rest of the system, and might not be desirable always.

Also once long document IDs have been inserted in the source DB. Simply deleting
them doesn't work as they'd still appear in the change feed. They'd have to
be purged or somehow skipped during the replication step. This commit helps
do the later.

Operators can configure maximum length via this setting:
```
  replicator.max_document_id_length=0
```

The default value is 0 which means there is no maximum enforced, which is
backwards compatible behavior.

During replication if maximum is hit by a document, that document is skipped,
an error is written to the log:

```
Replicator: document id `aaaaaaaaaaaaaaaaaaaaa...` from source db  `http://.../cdyno-0000001/` is too long, ignoring.
```

and `"doc_write_failures"` statistic is bumped.

COUCHDB-3291

nickva committed:
Allow configuring maximum document ID length during replication

Currently due to a bug in http parser and lack of document ID length
enforcement, large document IDs will break replication jobs. Large IDs
will pass through the _change feed, revs diffs,  but then fail
during open_revs get request. open_revs request will keep retrying until
it gives up after long enough time, then replication task crashes and
restart again with the same pattern. The current effective limit is
around 8k or so. (The buffer size default 8192 and if the first line
of the request is larger than that, request will fail).

(See http://erlang.org/pipermail/erlang-questions/2011-June/059567.html
for more information about the possible failure mechanism).

Bypassing the parser bug by increasing recbuf size, will alow replication
to finish, however that means simply spreading the abnormal document through
the rest of the system, and might not be desirable always.

Also once long document IDs have been inserted in the source DB. Simply deleting
them doesn't work as they'd still appear in the change feed. They'd have to
be purged or somehow skipped during the replication step. This commit helps
do the later.

Operators can configure maximum length via this setting:
```
  replicator.max_document_id_length=0
```

The default value is 0 which means there is no maximum enforced, which is
backwards compatible behavior.

During replication if maximum is hit by a document, that document is skipped,
an error is written to the log:

```
Replicator: document id `aaaaaaaaaaaaaaaaaaaaa...` from source db  `http://.../cdyno-0000001/` is too long, ignoring.
```

and `"doc_write_failures"` statistic is bumped.

COUCHDB-3291

eiri commented on src/couch_replicator_changes_reader.erl:
```diff
@@ -89,9 +89,19 @@ process_change(#doc_info{id = <<>>} = DocInfo, {_, Db, _, _, _}) ->
+        true ->
+            ShortId = lists:sublist(binary_to_list(Id), 64),
+            SourceDb = couch_replicator_api_wrap:db_uri(Db),
+            couch_log:error("Replicator: document id `~s...` from source db "
```
Just use `~64s...` in error format instead of allocating new variable, since you are not using it for anything else.

nickva replied:
Oh you're right. Well I should fix the primary PR as well. There is another config get change Robert suggested I will make that too.

nickva replied:
@eiri Since this is a cherry-picked commit of the upstream change. Would it make sense to merge this then have a separate PR for both upstream. Or maybe just have a commit for upstream which does this and then add that commit to this branch as well (just to make sure it doesn't diverge from upstream unnecessarily)

eiri replied:
I think simplest would be to get upstream ironed out and then update this branch accordingly, less movements, but if you really want to get this merged, say, to get it into next release, I'd be fine with that.

nickva replied:
upstream pr https://github.com/apache/couchdb-couch-replicator/pull/56

nickva committed:
Switch replicator max_document_id_length config to use infinity

Default value switched to be `infinity` instead of 0

COUCHDB-3291

nickva committed:
Use string formatting to shorten document ID during logging.

Previously used an explicit lists:sublist call but value was never used
anywhere besides the log message.

COUCHDB-3291

nickva commented:
@eiri updated with the cherry-picked commit from upstream

eiri reviewed (approved):
cool, +1

eiri commented:
just don't forget to squash before merging

eiri commented:
on a second thought, ignore the squashing bit and just try to keep things close to how they are on upstream.
